src/java/org/apache/cassandra/db/ReadCommand.java

```diff
@@ -436,15 +431,15 @@ public UnfilteredPartitionIterator executeLocally(ReadExecutionController execut
         long startTimeNanos = System.nanoTime();
 
         ColumnFamilyStore cfs = Keyspace.openAndGetStore(metadata());
-        Index index = getIndex(cfs);
+        Index.QueryPlan indexQueryPlan = indexQueryPlan();
 
         Index.Searcher searcher = null;
-        if (index != null)
+        if (indexQueryPlan != null)
         {
-            if (!cfs.indexManager.isIndexQueryable(index))
-                throw new IndexNotAvailableException(index);
+            cfs.indexManager.checkQueryability(indexQueryPlan);
 
-            searcher = index.searcherFor(this);
+            Index index = indexQueryPlan.getFirst();
+            searcher = indexQueryPlan.searcherFor(this);
             Tracing.trace("Executing read on {}.{} using index {}", cfs.metadata.keyspace, cfs.metadata.name, index.getIndexMetadata().name);
```

Comment:
nit: For now, this is probably the least disruptive change, given all active index groups are singletons, but at some point the trace message might need to include information about the plan not just "using index..."

Reply:
updated with suggested logs